<table style="width: 100%">
    <tr style="background: #ffffff">
        <td style="padding-top:25px;width: 180px"><img src="https://mci.edu/templates/mci/images/logo.svg" alt="Logo"></td>
        <td style="width: 100%">
            <div style="text-align:right; width: 100%; text-align:right"><font style="font-size:38px"><b>Grundlagen Programmierung</b></font></div>
            <div style="padding-top:0px; width: 100%; text-align:right"><font size="4"><b>MCI</b></font></div>
        </td>
    </tr>
</table>

---

# 2 Listen und Schleifen

Listen sind eine einfache Form Daten zu speichern. Kontrollstrukturen steuern den Programmablauf in imperativen Programmiersprachen. Im folgenden Notebook lernen Sie den Umgang mit Listen und for-Schleifen kennen. 

## Datentypen

Letztes Mal haben wir verschiedene Datentypen vorgestellt, die Variablen haben können.

In [5]:
type("ABC")

str

In [6]:
type(1)

int

In [7]:
type(1.1)

float

### Der Datentyp "None"

Es gibt immer wieder Situationen, in denen ein Eingabeparameter (etwa weil er optional ist) oder ein Rückgabewert (etwa weil ein Fehler aufgetreten ist) den Wert "None" annehmen soll. In Python gibt es dafür einen speziellen Datentyp `NoneType`, dessen Wertebereich genau einen Wert enthält, nämlich `None`.

None steht für die Abwesenheit eines Wertes. Dass die "Abwesenheit eines Wertes" durch einen speziellen Wert dargestellt wird (was eigentlich ein Widerspruch in sich ist), hat damit zu tun, dass der Computer mit "Nichts" nicht anders umgehen kann. Der Computer versteht nur Bitfolgen, und deshalb muss das "Nichts" durch eine spezielle Bitfolge dargestellt werden. Und in Python heißt diese spezielle Bitfolge `None`.

Wir wollen zum Beispiel eine neue Variable definieren, aber die Variable hat noch keinen Wert:

In [8]:
# Aktuelle Zeit in einer Simulation, die noch nicht gestartet wurde
current_time = None
type(current_time)

NoneType

### Boolesche Ausdrücke

Wie Sie gelernt haben, arbeiten Computer mit binären Ausdrücken (z.B. `True` oder `False`). Solche Informationen können in einem einzigen Bit gespeichert werden.  Ein boolescher Wert ist entweder `True` oder `False`. Boolesche Werte werden häufig verwendet, um bestimmte Bedingungen zu erfassen, z. B. ob eine Simulation gerade läuft oder ob ein Benutzer bestimmte Werte (z.B. die Start-Temperatur) bearbeiten kann:


In [9]:
simulation_active = True
can_edit = False

Wenn wir den Status einer booleschen Variable überprüfen wollen, können wir den Operator ```==``` verwenden. Wenn der Status von ```game_active``` tatsächlich ```True``` ist, ergibt die Auswertung einen anderen booleschen Wert ```True```. 

In [10]:

simulation_active = True  # Zuweisung, setzt den Wert
simulation_active == True # Vergleich 

True

Der aktuelle Wert von `simulation_active` ist nicht `False` deswegen gibt der folgende Vergleich `False` zurück:

In [11]:
simulation_active == False 

False

Vergleichende Auswertungen sind nicht auf boolesche Werte beschränkt. Wir können auch Zeichenketten oder Zahlen vergleichen:

In [12]:
"abc" == "abc"

True

In [13]:
"Abc" == "abc"

False

In [14]:
1 == 1

True

In [15]:
1 == 1.1

False

In [16]:
1 <= 1.1

True

✍️ **Aufgabe**

Schreiben Sie Code, der `simulation_active` und `current_time` zuerst auf `False` setzt und dann `simulation_active` auf `True` und `current_time` auf `0` und zuletzt überprüft, ob `current_time` immer auf `None` steht. 

In [17]:
#@title ❓ Click `Show code` in the code cell to show the solution. { display-mode: "form" }

simulation_active = False
current_time = None
simulation_active = True 
current_time = 0

current_time == None

False

### Boolean Algebra

Python kann Boolesche Ausdrücke mit den Schlüsselwörtern `and`, `or`, and `not` auswerten:

In [18]:
a = True
b = False

print("a and b is: " + str(a and b))

print("a or b is: " + str(a or b))

print("not a is: " + str(not(b)))

# we can also store the result in a new variable

c = (a or b) and a
print("c is: " + str(c))

a and b is: False
a or b is: True
not a is: True
c is: True


## Listen

In diesem und dem nächsten Kapitel lernen Sie was Listen sind und wie man mit den Elementen einer den Elementen einer Liste arbeitet. Mit Listen können Sie Informationen an einem Ort zu speichern, egal ob Sie nur ein paar Elemente oder Millionen von Elementen haben. Listen sind eine der mächtigsten Funktionalitäten in Python und verbinden viele wichtige Konzepte der Programmierung.


Letztes Mal haben wir eine Funktion geschrieben, die das Beschränkte Wachstum zu einem einzigen Zeitpunkt $t$ auswertet.
Wenn wir mehrere Zeitpunkte speichern wollen, an denen wir die Ergebnisse auswerten wollen, könnten wir mehrere Variablen verwenden

```
t_in_minutes_2 = 5
t_in_minutes_2 = 10
t_in_minutes_2 = 15
t_in_minutes_2 = 20
...
```

Listen sind eine elegantere Möglichkeit, dies zu tun. Unten finden Sie noch einmal die Implementierung des beschränkten Wachstums. Wenn wir die Temperaturentwicklung über die Zeit ausrechnen wollten, könnten wir dies auch ohne eine Liste oder Schleifer tun, indem wir die Werte der Variablen `t_in_minutes` immer wieder anpassen:


```
# importing the math library
import math

# here the user can define the input
initial_temperature = 0
maximum_temperature = 30
growth_rate = 0.02

t_in_minutes = 10


# here the calculation is made
current_temperature = maximum_temperature - (maximum_temperature - initial_temperature)   * math.exp(- growth_rate*t_in_minutes)
current_temperature
```

#### Was ist eine Liste?

Eine Liste ist eine Sammlung von Elementen in einer bestimmten Reihenfolge. Sie können eine Liste erstellen, die die Buchstaben des Alphabets, die Ziffern von 0-9 oder die Namen aller Personen in Ihrer Familie enthält. Sie können alles in eine Liste schreiben, was Sie wollen, und die Elemente in Ihrer Liste müssen nicht in einem bestimmten Zusammenhang stehen.

Da eine Liste normalerweise mehr als ein Element enthält, ist es eine gute Idee, den Namen der Liste im Plural zu schreiben, wie `letters`, `digits` oder `names`.

In Python zeigen eckige Klammern (``[]``) eine Liste an, und einzelne Elemente in der Liste werden durch Kommas getrennt. Hier ist ein einfaches Beispiel für eine Liste, die einige Arten von Heizungstypen enthält:

In [19]:
heizungstypen = ['infrarot', 'wärmepumpe', 'öl', 'gas']
print(heizungstypen)

['infrarot', 'wärmepumpe', 'öl', 'gas']


In [20]:
type(heizungstypen)

list

#### Zugriff auf Elemente in einer Liste

Listen sind geordnete Sammlungen, so dass Sie auf jedes Element in einer Liste zugreifen können, indem Sie Python die Position oder den ``index`` des gewünschten Elements mitteilen. Um auf ein Element in einer Liste zuzugreifen, schreiben Sie den Namen der Liste gefolgt vom Index des Elements in eckigen Klammern.
Ein Beispiel: Ziehen wir das erste Fahrrad aus der Liste Fahrräder heraus:


In [21]:
print(heizungstypen[0])

infrarot


##### Index Positions start at `0`, Not `1` 

Python betrachtet das erste Element in einer Liste als Position 0, nicht als Position 1. Dies gilt für die meisten Programmiersprachen, und der Grund hat damit zu tun, wie die Listenoperationen auf einer niedrigeren Ebene implementiert sind. 
Wenn Sie unerwartete Ergebnisse erhalten, kann dies häufig an einem einfachen Off-by-One-Fehler liegen.

In [22]:
heizungstypen = ['infrarot', 'wärmepumpe', 'öl', 'gas']

print(heizungstypen[1])
print(heizungstypen[3])
print(heizungstypen[4])

wärmepumpe
gas


IndexError: list index out of range

Python hat eine spezielle Syntax für den Zugriff auf das letzte Element in einer Liste. Wenn Sie nach dem Element bei Index `-1` fragen, gibt Python immer das letzte Element in der Liste zurück:

In [ ]:
heizungstypen = ['infrarot', 'wärmepumpe', 'öl', 'gas']

print(heizungstypen[-1])

gas


## Besprechung der Aufgaben

- Sind Sie mit ein Aufgaben bis hierhin fertig geworden und haben Sie keine Fragen, so legen Sie die grüne Karte nach oben
- Sind Sie mit den Aufgaben fertig, aber hätten gerne, dass die Lösungen nocheinmal gemeinsam besprochen werden, legen Sie die gelbe Karte nach oben

![](https://www.lokalinfo.ch/fileadmin/news_import/image003_03.jpg)

✍️ **Aufgabe**

Die Entwickeln eine Smart Home App, die z.B. Wärmepräferenzen verschiedener Nutzer:innen speichern kann. Erstellen Sie eine Liste `persons` mit den Namen von vier Personen.
Ihre App soll jede Person individuell mit einer Nachricht begrüßen können.

`"Hallo Julian, willkommen zurück!"`

Geben Sie diese Nachricht aus und nutzen Sie dazu die Position des Namens in der Liste, wobei sie die Position in der Variable `person_number` speichern. 

*Hinweis:* Erinnern Sie sich an den `"String {}".format(<variable>)` Befehl aus dem letzten Notebook.


In [ ]:
# Anlegen der Liste

# Festlegen der person_number

# Print Statement

In [ ]:
#@title ❓ Click `Show code` in the code cell to show the solution. { display-mode: "form" }

persons = ["Julian", "Andreas", "Philipp", "Manuel"]

person_number = 1

print("Hallo {}, willkommen zurück!".format(persons[person_number]))

Hallo Andreas, willkommen zurück!


#### Hinzufügen und Entfernen von Elementen zu/von einer Liste

Python bietet mehrere Möglichkeiten, neue Daten zu bestehenden Listen hinzuzufügen.

**Anhängen/Append**

Anhängen von Elementen an das Ende einer Liste Die einfachste Art, ein neues Element zu einer Liste hinzuzufügen, ist das Anhängen des Elements an die Liste. Wenn Sie ein Element an eine Liste anhängen, wird das neue Element an das Ende der Liste angefügt, um beispielsweise eine neue Person anzuhängen.

In [ ]:
persons = ["Julian", "Andreas", "Philipp", "Manuel"]
print(persons)
persons.append('Tim')
print(persons)

['Julian', 'Andreas', 'Philipp', 'Manuel']
['Julian', 'Andreas', 'Philipp', 'Manuel', 'Tim']


Die `append()`-Methode macht es einfach, Listen dynamisch aufzubauen. Sie können zum Beispiel mit einer leeren Liste beginnen und der Liste dann Elemente hinzufügen mit einer Reihe von append()-Anweisungen hinzufügen.

In [ ]:
persons = []
persons.append('Julian')
persons.append('Andreas')
persons.append('Philipp')
print(persons)

['Julian', 'Andreas', 'Philipp']


**Entfernen eines Elements nach Wert**

Manchmal kennt man die Position des Wertes, den man aus einer Liste entfernen möchte. Wenn Sie nur den Wert des Elements kennen, das Sie entfernen möchten, können Sie die Methode ```remove()``` verwenden.

Zum Beispiel wollen wir Experimente mit verschiedenen Heizungsanalgen planen. Das Experiment am 21.03.2023 entfällt und wird deswegen aus dem Stapel entfernt.

In [ ]:
lab_dates = ['21.03.2023', '22.03.2023', '24.03.2023']
print(lab_dates)
failed_experiments = '21.03.2023'
lab_dates.remove(failed_experiments)
print(lab_dates)
print(failed_experiments)

print("\n Experiments am " + failed_experiments + " entfernt.")



['21.03.2023', '22.03.2023', '24.03.2023']
['22.03.2023', '24.03.2023']
21.03.2023

 Experiments am 21.03.2023 entfernt.


✍️ **Aufgabe**

Fügen Sie ein Experiment am `'25.03.2023'` zur Liste `lab_dates` hinzu und entfernen Sie das am `'24.03.2023'`.

In [ ]:
#@title ❓ Click `Show code` in the code cell to show the solution. { display-mode: "form" }

lab_dates = ['21.03.2023', '22.03.2023', '24.03.2023']
lab_dates.append('25.03.2023')
lab_dates.remove('24.03.2023')
print(lab_dates)

['21.03.2023', '22.03.2023', '25.03.2023']


#### Ordnen einer Liste

Oft werden Ihre Listen in einer unvorhersehbaren Reihenfolge erstellt, weil Sie nicht immer kontrollieren können in welcher Reihenfolge die Benutzer ihre Daten eingeben. Obwohl dies in den meisten Fällen unvermeidbar ist, werden Sie Ihre Informationen häufig in einer Informationen in einer bestimmten Reihenfolge präsentieren. Manchmal möchten Sie die ursprüngliche Reihenfolge Ihrer Liste beibehalten, und ein anderes Mal wollen Sie die ursprüngliche
Reihenfolge ändern. Python bietet eine Reihe von Möglichkeiten, Ihre Listen zu organisieren, abhängig von der jeweiligen Situation.



##### Dauerhaftes Sortieren einer Liste mit der Methode `sort()`

Die Methode  `sort()` operiert auf der Liste und verändert diese. Wie erkennen das, daran, dass wir das Schlüsselwort `.sort()` hinten anhängen und somit auf die Liste aufrufen. 

🤓 *Hinweis:* Wer sich schon einmal mit Objektorientierter Programmierung auseinander gesetzt hat, erkennt, dass hier eine Methode `sort()` auf eines Objektes der Klasse `Liste` aufgerufen wird. For heute reicht es sich zu merken, dass die Liste so dauerhaft sortiert wird.

In [ ]:
heizungstypen = ['infrarot', 'wärmepumpe', 'öl', 'gas']
heizungstypen.sort()
print(heizungstypen)

['audi', 'bmw', 'subaru', 'toyota']


Wir können dem `sort()`-Befehl auch bestimmte Parameter übergeben, um z.B. andersherum zu sortieren.

In [ ]:
heizungstypen = ['infrarot', 'wärmepumpe', 'öl', 'gas']
heizungstypen.sort(reverse=True)
print(heizungstypen)

['toyota', 'subaru', 'bmw', 'audi']


Wenn wir die Liste ausgeben oder printen, wird diese zurückgegeben.
Wenn wir allerdings versuchen die die Funktion zu printen sehen wird, dass diese keinen Typ bzw. keine Rückgabe (Rückgabe-Typ `None`) hat. Wir können das Ergebnis also nicht in einer neuen Variable speichern. Allerdings haben wir das sortierte Ergebnis ja in unser ursprünglichen Liste `cars` und können damit weiterarbeiten.

In [ ]:
my_sorted_list = heizungstypen.sort(reverse=True)
print(type(my_sorted_list))

<class 'NoneType'>


In [ ]:
heizungstypen

['toyota', 'subaru', 'bmw', 'audi']

#### Ermitteln der Länge einer Liste

Mit der Funktion `len()` können Sie schnell die Länge einer Liste ermitteln. Die Liste in diesem Beispiel hat vier Elemente, ihre Länge ist also 4:

In [ ]:
heizungstypen = ['infrarot', 'wärmepumpe', 'öl', 'gas']
len(heizungstypen)

4

#### Vermeidung von Indexfehlern bei der Arbeit mit Listen

Beachten Sie, dass der Index von 0 bis 3 reicht!

In [ ]:
heizungstypen[4]

IndexError: list index out of range

Denken Sie daran, dass Sie immer dann, wenn Sie auf das letzte Element in einer Liste zugreifen wollen
verwenden Sie den Index -1. Dies funktioniert immer, auch wenn sich die Liste seit dem letzten Zugriff geändert hat
Größe geändert hat, seit Sie das letzte Mal auf sie zugegriffen haben:

In [ ]:
heizungstypen[-1]

'subaru'

In [ ]:
heizungstypen[len(heizungstypen)-1]

'subaru'

## Besprechung der Aufgaben

- Sind Sie mit ein Aufgaben bis hierhin fertig geworden und haben Sie keine Fragen, so legen Sie die grüne Karte nach oben
- Sind Sie mit den Aufgaben fertig, aber hätten gerne, dass die Lösungen nocheinmal gemeinsam besprochen werden, legen Sie die gelbe Karte nach oben

![](https://www.lokalinfo.ch/fileadmin/news_import/image003_03.jpg)

## Arbeiten mit Listen

Sie können nun Listen erstellen und mit den einzelnen Elementen einer Liste zu arbeiten. In diesem Kapitel lernen Sie, wie Sie mit wenigen Zeilen Code durch eine ganze Liste durchlaufen, egal wie lang die Liste ist. Mit einer Schleife können Sie die gleiche Aktion oder eine Reihe von Aktionen durchführen,
mit jedem Element in einer Liste durchführen. So können Sie effizient arbeiten mit Listen beliebiger Länge arbeiten, einschließlich solcher mit Tausenden oder sogar Millionen von Elementen.


### Schleifen über eine ganze Liste

Oft möchte man alle Einträge in einer Liste durchlaufen und mit jedem Eintrag die gleiche Aufgabe mit jedem Eintrag durchführen. In einem CAD-Programm möchten Sie zum Beispiel jedes Element auf dem Bildschirm um denselben Betrag verschieben.
Element auf dem Bildschirm um denselben Betrag verschieben, oder in einer Liste mit Zahlen vielleicht dieselbe statistische Operation für jedes Element durchführen. Oder vielleicht möchten Sie jede Überschrift aus einer Liste von Artikeln auf einer Website anzeigen lassen.
Wenn Sie die gleiche Aktion mit jedem Element in einer Liste durchführen wollen, können Sie die for-Schleife von Python verwenden.

Stellen Sie sich vor, Sie haben eine Liste mit allen Terminen, an denen Sie `n` Experimente im Labor gemacht haben. Mit dem Befehl ``for`` können Sie eine Schleife durch die Liste ziehen.

- Wir beginnen damit, eine Liste zu definieren, wie wir es zuvor getan haben. Als nächstes, definieren wir eine **for-Schleife**. Diese Zeile weist Python an, ein Datum aus der Liste `lab_dates` zu ziehen und es in der Variablen `lab_date` zu speichern. Dann sagen wir Python, dass es das Datum ausgeben soll, das gerade in `lab_date` gespeichert wurde. Python wiederholt dann die Zeilen 2 und 3, einmal für jedes Datum in der Liste.

In [ ]:
lab_dates = ['21.03.2023', '22.03.2023', '24.03.2023']
for lab_date in lab_dates:
  print("There was a lab on {}.".format(lab_date))

There was a lab on 21.03.2023.
There was a lab on 22.03.2023.
There was a lab on 24.03.2023.


Wenn Sie zum ersten Mal Schleifen verwenden, denken Sie daran, dass der Satz von
Schritten für jedes Element in der Liste einmal wiederholt wird, unabhängig davon, wie viele Elemente in der Liste sind. Wenn Sie eine Million Einträge in Ihrer Liste haben, wiederholt Python diese Schritte eine Million Mal - und normalerweise sehr schnell.

Denken Sie beim Schreiben Ihrer eigenen for-Schleifen auch daran, dass Sie
Namen für die temporäre Variable, die jeden Wert in der Liste enthält, frei wählen können.
Es ist jedoch hilfreich, einen aussagekräftigen Namen zu wählen, der für ein
einzelnes Element der Liste darstellt. So können Sie beispielsweise eine for-Schleife wie folgt beginnen für eine Liste von Katzen, eine Liste von Hunden und eine allgemeine Liste von Elementen:

```
for lab_date in lab_dates:
for cat in cats:
for dog in dogs:
for item in list_of_items:
```

Umfang der Schleife

Python verwendet Einrückungen (Leerzeichen), um anzuzeigen, dass eine Zeile Code Teil der Schleife ist. Die ersten beiden `print()`-Anweisungen stehen innerhalb der Schleife. Die letzte Anweisung liegt außerhalb der Schleife und wird daher nur einmal am Ende ausgeführt.

Eine Einrückung in Python besteht normalerweise aus vier Leerzeichen (`"    "`), der schnellere Weg eine Einrückung zu erzeugen ist die Tabulator-Taste (↹)


In [ ]:
lab_dates = ['21.03.2023', '22.03.2023', '24.03.2023']
for lab_date in lab_dates:
  print("There was a lab on {}.".format(lab_date))
  print("The lab was a great success!")
print("There were {} labs in total.".format(len(lab_dates)))

There was a lab on 21.03.2023.
The lab was a great success!
There was a lab on 22.03.2023.
The lab was a great success!
There was a lab on 24.03.2023.
The lab was a great success!
There were 3 labs in total.


✍️ **Aufgabe**

Kopieren Sie den Code von oben und fügen Sie die Zeile `print("... \n")` so ein, dass Sie jedes Mal nach `"The lab was a great success!"` erscheint.

In [ ]:
#@title ❓ Click `Show code` in the code cell to show the solution. { display-mode: "form" }

lab_dates = ['21.03.2023', '22.03.2023', '24.03.2023']
for lab_date in lab_dates:
  print("There was a lab on {}.".format(lab_date))
  print("The lab was a great success!")
  print("... \n")
print("There were {} labs in total.".format(len(lab_dates)))

There was a lab on 21.03.2023.
The lab was a great success!
... 

There was a lab on 22.03.2023.
The lab was a great success!
... 

There was a lab on 24.03.2023.
The lab was a great success!
... 

There were 3 labs in total.


### Fehler bei der Einrückung vermeiden

Python verwendet Einrückungen, um festzustellen, wann eine Codezeile mit der darüber liegenden Zeile verbunden ist.
Die Einrückung in Python macht den Code sehr einfach zu lesen. Im Grunde verwendet Python Leerraum, um Sie zu zwingen, sauber formatierten Code mit einer klaren visuellen Struktur zu schreiben. In längeren Python-Programmen werden Sie feststellen, dass Codeblöcke auf verschiedenen Ebenen eingerückt sind. Diese Einrückungsebenen helfen Ihnen, einen Überblick über die Gesamtorganisation des Programms zu bekommen.

Wenn Sie anfangen, Code zu schreiben, der auf die richtige Einrückung angewiesen ist, müssen Sie
müssen Sie auf einige häufige Einrückungsfehler achten.

In [ ]:
magicians = ['alice', 'david', 'carolina']
for magician in magicians:
print(magician)

IndentationError: ignored

Sie können diese Art von Einrückungsfehler in der Regel beheben, indem Sie die
Zeile oder Zeilen unmittelbar nach der for-Anweisung einrücken.

Wenn Sie versehentlich eine Zeile einrücken, die nicht eingerückt werden muss, informiert Python Sie über den unerwarteten Einzug:

In [ ]:
message = "Hello Python world!"
  print(message)

IndentationError: ignored

### Numerische Listen erstellen

Es gibt viele Gründe, eine Reihe von Zahlen zu speichern. Zum Beispiel müssen Sie die Positionen der einzelnen Spielfiguren in einem Spiel aufzeichnen, und Sie möchten vielleicht auch die Highscores der Spieler speichern. Bei Datenvisualisierungen werden Sie fast immer mit Zahlenreihen arbeiten, z. B. Temperaturen, Entfernungen, Bevölkerungsgrößen oder Längen- und Breitengrade, neben anderen Arten von numerischen Mengen.
Listen sind ideal zum Speichern von Zahlenmengen, und Python bietet eine Reihe von Werkzeuge zur Verfügung, die Ihnen helfen, effizient mit Listen von Zahlen zu arbeiten. Sobald Sie verstanden haben, wie man diese Werkzeuge effektiv einsetzt, wird Ihr Code auch dann gut funktionieren wenn Ihre Listen Millionen von Elementen enthalten.

#### Verwendung der Funktion range()
Pythons Funktion `range()` macht es einfach, eine Reihe von Zahlen zu erzeugen.
Zum Beispiel können Sie die Funktion `range()` verwenden, um eine Reihe von Zahlen zu drucken wie hier:

In [ ]:
for value in range(1,5):
  print(value)

1
2
3
4


Obwohl dieser Code so aussieht, als sollte er die Zahlen von 1 bis 5 ausgeben, wird die Zahl 5 nicht ausgegeben.
Wenn Ihre Ausgabe anders ist als das, was Sie erwarten, wenn Sie die Funktion
`range()` verwenden, versuchen Sie, den Endwert um 1 zu erhöhen.


#### Mit range() eine Liste von Zahlen erstellen

Wie Sie in der nächsten Zeile sehen, ist ein Bereich ein spezieller Datentyp in Python. Wie Listen können wir auch Bereiche in einer for-Schleife verwenden (Siehe Beispiel zuvor). 

In [ ]:
type(range(1,5))

range

Mit `range()` kann aber auch eine Liste von Zahlen erstellt werden, wenn Sie die Range mit der Funtion ``list()`` in eine Liste umwandeln (dies ist ähnlich zum Prinzip das Castings, das wir bereits angewendet haben). Wenn Sie `list()` um einen Aufruf der Funktion `range()` einschließen, wird die Ausgabe eine Liste von Zahlen sein.

In [ ]:
numbers = list(range(1,6))
print(numbers)

[1, 2, 3, 4, 5]


Wir können auch die Funktion `range()` verwenden, um Python anzuweisen, die Zahlen in einem bestimmten Bereich zu überspringen.
Gewöhnen Sie sich an die [Dokumentation](https://docs.python.org/3/library/stdtypes.html#range) von Funktionen zu lesen. Hier stehen häufig Informationen, die Ihnen das Schreiben von Code erleichtern.
So würden wir zum Beispiel die geraden Zahlen auflisten zwischen 1 und 10 auflisten:

In [ ]:
even_numbers = list(range(2,11,2))
print(even_numbers)

[2, 4, 6, 8, 10]


✍️ **Aufgabe**

Erstellen Sie eine Liste `liste` mit den die alle 5. Minuten in zwei Stunden enthält, als `[0,5,10,...]`. Stellen Sie sicher, dass Sie am Ende keine `range` sondern eine richtige `list` haben. So eine Liste können wir später verwenden, wenn wir automatisiert das Beschränkte Wachstum zu verschiedenen Zeitpunkten bestimmen wollen.

In [ ]:
#@title ❓ Click `Show code` in the code cell to show the solution. { display-mode: "form" }
liste = list(range(0,121,5))
print(liste)
print(type(liste))

[0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95, 100, 105, 110, 115, 120]
<class 'list'>


## Besprechung der Aufgaben

- Sind Sie mit ein Aufgaben bis hierhin fertig geworden und haben Sie keine Fragen, so legen Sie die grüne Karte nach oben
- Sind Sie mit den Aufgaben fertig, aber hätten gerne, dass die Lösungen nocheinmal gemeinsam besprochen werden, legen Sie die gelbe Karte nach oben

![](https://www.lokalinfo.ch/fileadmin/news_import/image003_03.jpg)

🏆 **Aufgabe**

Fügen Sie den Code aus dem Notebook zum beschränkten Wachstum in eine for-Schleife ein, um die Ergebnisse von $t=0$ bis $t=120$ in Schritten von $5$ zu erhalten. Nutzen Sie dafür eine Range oder eine Liste, die Sie aus der Range erstellen.

Passen Sie auch den String an, so dass die Simulationsparameter zu Beginn der Simulation nur einmal erwähnt werden.

*Hinweis*: Überlegen sie zunächst, welche Bereiche des Codes in der Schleife müssen (also wiederholt werden) und welche nicht.

```
# importing the math library
import math

# here the user can define the input
initial_temperature = 0
maximum_temperature = 30
growth_rate = 0.02

t_in_minutes = 10


# here the caclulation is made
current_temperature = maximum_temperature - (maximum_temperature - initial_temperature)   * math.exp(- growth_rate*t_in_minutes)


# here, we print out the results
print("Started simulation with the following parameters: \n inital temperature: {} \n maximum temperature: {} \n growth rate: {}".format(initial_temperature, maximum_temperature, growth_rate)) 

print("The temperature after {0} minutes is {1:.2f} °C".format(t_in_minutes, current_temperature))
```



In [ ]:
#@title ❓ Click `Show code` in the code cell to show the solution. { display-mode: "form" }


# importing the math library
import math

# here the user can define the input
initial_temperature = 0
maximum_temperature = 30
growth_rate = 0.02

t_in_minutes = 10

print("Started simulation with the following parameters: \n initial temperature: {} \n maximum temperature: {} \n growth rate: {}".format(initial_temperature, maximum_temperature, growth_rate)) 

for t_in_minutes in range(0,121,5):
  current_temperature = maximum_temperature - (maximum_temperature - initial_temperature)   * math.exp(- growth_rate*t_in_minutes)


  print("The temperature after {0} minutes is {1:.2f} °C".format(t_in_minutes, current_temperature))



Started simulation with the following parameters: 
 initial temperature: 0 
 maximum temperature: 30 
 growth rate: 0.02
The temperature after 0 minutes is 0.00 °C
The temperature after 5 minutes is 2.85 °C
The temperature after 10 minutes is 5.44 °C
The temperature after 15 minutes is 7.78 °C
The temperature after 20 minutes is 9.89 °C
The temperature after 25 minutes is 11.80 °C
The temperature after 30 minutes is 13.54 °C
The temperature after 35 minutes is 15.10 °C
The temperature after 40 minutes is 16.52 °C
The temperature after 45 minutes is 17.80 °C
The temperature after 50 minutes is 18.96 °C
The temperature after 55 minutes is 20.01 °C
The temperature after 60 minutes is 20.96 °C
The temperature after 65 minutes is 21.82 °C
The temperature after 70 minutes is 22.60 °C
The temperature after 75 minutes is 23.31 °C
The temperature after 80 minutes is 23.94 °C
The temperature after 85 minutes is 24.52 °C
The temperature after 90 minutes is 25.04 °C
The temperature after 95 minute

## Besprechung der Aufgaben

- Sind Sie mit ein Aufgaben bis hierhin fertig geworden und haben Sie keine Fragen, so legen Sie die grüne Karte nach oben
- Sind Sie mit den Aufgaben fertig, aber hätten gerne, dass die Lösungen nocheinmal gemeinsam besprochen werden, legen Sie die gelbe Karte nach oben

![](https://www.lokalinfo.ch/fileadmin/news_import/image003_03.jpg)

#### Einfache Statistik mit einer Liste von Zahlen

Einige Python-Funktionen sind spezifisch für Listen von Zahlen. Zum Beispiel können Sie auf einfache Weise das Minimum, Maximum und die Summe einer Liste von Zahlen ermitteln:

In [ ]:
digits = [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
min(digits)

0

In [ ]:
max(digits)

9

In [ ]:
sum(digits)

45

✍️  **Aufgabe**

Berechnen Sie den Mittelwert (Durchschnitt) der Liste `digits` indem Sie die Funktionen `sum()` und `len()` nutzen. 

In [ ]:
#@title ❓ Click `Show code` in the code cell to show the solution. { display-mode: "form" }

mean =  sum(digits) / len(digits)
mean

4.5

✍️ **Aufgabe**

Eine wichtiges Prinzip ist es Code mit einfachen Beispielen zu testen

- Zählen bis zwanzig: Verwenden Sie eine for-Schleife, um die Zahlen von 1 bis einschließlich 20 auszugeben und speichern Sie diese in einer Liste `numbers`
- Initialisieren Sie dann vor der Schleife `numbers` als leere Liste
- Verwenden Sie dann `min()` und `max()`, um sicherzustellen, dass Ihre Liste tatsächlich bei eins beginnt und bei 20 endet
- Verwenden Sie auch die Funktion `sum()`, um alle Zahlen in Ihrer Liste zusammen zu zählen

- Einhundert: Führen Sie Ihren Code für die Zahlen von 1 bis 100 aus, um die [Gaußsche Summenformel](https://www.giga.de/ratgeber/specials/gausssche-summenformel-zahlen-von-1-bis-100-addieren-so-geht-s/) zu überprüfen
- Eine Million: Führen Sie ihren Code auch hiermit aus, aber kommentieren Sie zuvor Ihr `print()`-statement aus!

In [ ]:
for value in range(1,21):
  print(value)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20


In [ ]:
#@title ❓ Click `Show code` in the code cell to show the solution. { display-mode: "form" }

numbers = []
for value in range(1,100+1):
  numbers.append(value)
print("Minimum is: {}".format(min(numbers)))
print("Maximum is: {}".format(max(numbers)))
print("Sum is: {}".format(sum(numbers)))
print("Mean is: {}".format(sum(numbers)/len(numbers)))

Minimum is: 1
Maximum is: 100
Sum is: 5050
Mean is: 50.5


### Arbeiten mit einem Teil einer Liste

![](https://media.geeksforgeeks.org/wp-content/uploads/20201025141338/Screenshot584.png)

#### Aufteilen einer Liste

Um ein Slice zu erstellen, geben Sie den Index des ersten und letzten Elements an, mit dem Sie arbeiten wollen. Wie bei der Funktion `range()` stoppt Python ein Element vor dem zweiten Index, den Sie angeben. Zur Ausgabe der ersten drei Elemente, würden Sie die Indizes 0 bis 3 anfordern, was die Elemente
0, 1 und 2.

Im folgenden Beispiel geht es um eine Liste von Spielern einer Mannschaft:

In [ ]:
players = ['charles', 'martina', 'michael', 'florence', 'eli']
print(players[0:3])

['charles', 'martina', 'michael']


Sie können eine beliebige Teilmenge einer Liste erzeugen. Wenn Sie zum Beispiel das zweite, dritte und vierte Element einer Liste benötigen, beginnen Sie das Slice bei Index 1 und bei Index 4 enden:

In [ ]:
print(players[1:4])

['martina', 'michael', 'florence']


Wenn Sie den ersten Index in einem Slice weglassen, beginnt Python automatisch am Anfang der Liste:

In [ ]:
print(players[:4])

['charles', 'martina', 'michael', 'florence']


Eine ähnliche Syntax funktioniert, wenn Sie einen Ausschnitt wünschen, der das Ende einer Liste umfasst.
Wenn Sie zum Beispiel alle Elemente vom dritten bis zum letzten Element haben wollen,
können Sie mit Index 2 beginnen und den zweiten Index weglassen:

In [ ]:
print(players[2:])

['michael', 'florence', 'eli']


Mit dieser Syntax können Sie alle Elemente von einem beliebigen Punkt in Ihrer Liste bis zum Ende ausgeben, unabhängig von der Länge der Liste. Erinnern Sie sich, dass ein negativer Index ein Element in einer bestimmten Entfernung vom Ende der Liste ausgibt;
Daher können Sie jedes beliebige Stück vom Ende einer Liste ausgeben. Zum Beispiel, wenn wir die letzten drei Spieler auf der Liste ausgeben wollen:

In [ ]:
print(players[-3:])

['michael', 'florence', 'eli']


✍️ **Aufgabe**

Geben sie die letzten beiden Spieler aus. Nutzen Sie einmal die Notation mit den direkten Index-Werten und einmal die Notation mit dem  `-` für die letzten beiden. 

In [ ]:
#@title ❓ Click `Show code` in the code cell to show the solution. { display-mode: "form" }
print(players[-2:])
print(players[3:5])

['florence', 'eli']
['florence', 'eli']


### Kopieren einer Liste

Oft möchten Sie mit einer bestehenden Liste beginnen und eine völlig neue Liste erstellen die auf der ersten Liste basiert. Wir wollen uns ansehen, wie das Kopieren einer Liste funktioniert und eine Situation, in der das Kopieren einer Liste nützlich ist.
Um eine Liste zu kopieren, können Sie ein Slice erstellen, das die gesamte ursprüngliche Liste umfasst indem man den ersten und den zweiten Index (`[:]`) weglässt. Dadurch wird Python angewiesen, ein ein Slice zu erstellen, das mit dem ersten Element beginnt und mit dem letzten Element endet, wodurch eine
eine Kopie der gesamten Liste.

Nehmen wir an, wir haben eine Liste unserer Lieblingsspeisen und wollen
eine separate Liste der Lebensmittel erstellen, die ein Freund mag. Dieser Freund mag alles in unserer Liste, also können wir seine Liste durch Kopieren unserer Liste erstellen:

In [ ]:
my_foods = ['pizza', 'falafel', 'carrot cake']
friend_foods = my_foods[:]
print("My favorite foods are:")
print(my_foods)
print("\nMy friend's favorite foods are:")
print(friend_foods)

My favorite foods are:
['pizza', 'falafel', 'carrot cake']

My friend's favorite foods are:
['pizza', 'falafel', 'carrot cake']


Um zu beweisen, dass wir tatsächlich zwei getrennte Listen haben, fügen wir in jede Liste ein neues Lebensmittel ein und zeigen, dass jede Liste die Lieblingsspeisen der entsprechenden Person aufzeichnet:

In [ ]:
my_foods = ['pizza', 'falafel', 'carrot cake']
friend_foods = my_foods[:]
my_foods.append('cannoli')
friend_foods.append('ice cream')

print("My favorite foods are:")
print(my_foods)

print("\nMy friend's favorite foods are:")
print(friend_foods)

My favorite foods are:
['pizza', 'falafel', 'carrot cake', 'cannoli']

My friend's favorite foods are:
['pizza', 'falafel', 'carrot cake', 'ice cream']


Die Ausgabe zeigt, dass 'Cannoli' nun in unserer Liste der Lieblingsspeisen erscheint, aber 'Eiscreme' nicht. 
Bei unserem Freund können wir sehen, dass 'Eiscreme' in der Liste auftaucht, aber 'Cannoli' nicht. 

Hätten wir einfach `friend_foods` gleich `my_foods` gesetzt, würden wir nicht zwei getrennte Listen erzeugen. 

Hier sehen Sie, was passiert, wenn Sie versuchen, eine Liste zu kopieren, ohne ein Slice zu verwenden:

In [ ]:
my_foods = ['pizza', 'falafel', 'carrot cake']

# This is dangerous:
friend_foods = my_foods

my_foods.append('cannoli')
friend_foods.append('ice cream')
print("My favorite foods are:")
print(my_foods)
print("\nMy friend's favorite foods are:")
print(friend_foods)

My favorite foods are:
['pizza', 'falafel', 'carrot cake', 'cannoli', 'ice cream']

My friend's favorite foods are:
['pizza', 'falafel', 'carrot cake', 'cannoli', 'ice cream']


Anstatt eine Kopie von `my_foods` in `friend_foods` zu speichern, setzen wir
`friend_foods` gleich `my_foods`. Diese Syntax sagt Python, dass es die neue Variable `friend_foods` mit der bestehden Liste `my_foods` zu verbinden, so dass nun beide Variablen auf dieselbe Liste (Stelle im Speicher) verweisen. 
Das Ergebnis ist, dass, wenn wir "Cannoli" zu `my_foods` hinzufügen, dies auch in `friend_foods` erscheint. Ebenso taucht "Eiscreme" in beiden Listen auf, auch wenn sie nur in `friend_foods` hinzugefügt wird.


Dass wird aus versehen ein Objekt verändern, das wir eigentlich kopieren wollten, ist ein häufiger Fehler.

Das liegt daran, dass wird es gewohnt sind das dieses Verhalten bei anderen Datentypen, wie Zahlen und Zeichenketten nicht auftritt. Dies liegt draran, das Elementare Datentypen wie Integers nicht veränderbar sind und sich die Zuweisung [deswegen anders verhält](https://www.codingem.com/python-copy-list/).

![](https://www.codingem.com/wp-content/uploads/2022/02/reference.jpeg)

In der Praxis, sollte man sich angewöhnen. Listen und andere komplexe Datenobjekte zu kopieren und nicht über eine Gleichsetzung zu definieren!

In [ ]:
# Mit primitiven Datentypen ist dies jedoch kein Problem:
a = 10
b = a
a = 50
print(a, b)

50 10


## Besprechung der Aufgaben

- Sind Sie mit ein Aufgaben bis hierhin fertig geworden und haben Sie keine Fragen, so legen Sie die grüne Karte nach oben
- Sind Sie mit den Aufgaben fertig, aber hätten gerne, dass die Lösungen nocheinmal gemeinsam besprochen werden, legen Sie die gelbe Karte nach oben

![](https://www.lokalinfo.ch/fileadmin/news_import/image003_03.jpg)

## Zusammenfassung

Mit Listen haben Sie nun die wichtigste Datenstruktur kennengelernt, mit denen mehre einzelne Daten in einem Datensatz zusammengefasst werden können. Sie sind nun in der Lage Listen anzulegen und manipulieren. Außerdem haben Sie gelernt Aufgaben für eine gesamte Liste durchzuführen und so zu automatisieren. 




---

# 🤓Optional: Folgende Aufgaben sind einige nützliche Python-Funktionen für Fortgeschrittene, die Sie nicht für den weiteren Kurs brauchen

##### 🤓 Vorübergehendes Sortieren einer Liste 

Mit der Funktion `sorted()` hingegen können wir die Reihenfolge einer ursprünglichen Liste beizubehalten, sie aber in einer sortierten Reihenfolge darstellen und in einer neuen Variable speichern. verwenden. Die Funktion `sorted()` übernimmt eine ursprüngliche Liste als Parameter und gibt eine sortierte Liste zurück. Die originale Liste wird dabei nicht verändert.

In [ ]:
print("Here is the original list:")
print(heizungstypen)
print("\nHere is the sorted list:")
print(sorted(heizungstypen))
print("\nHere is the original list again:")
print(heizungstypen)

Here is the original list:
['toyota', 'subaru', 'bmw', 'audi']

Here is the sorted list:
['audi', 'bmw', 'subaru', 'toyota']

Here is the original list again:
['toyota', 'subaru', 'bmw', 'audi']


🤓 **Removing**

Entfernen eines Elements mit der Anweisung `del`
Wenn Sie die Position des Elements kennen, das Sie aus einer Liste entfernen möchten, können Sie die `del`-Anweisung verwenden.

In [ ]:
motorcycles = ['honda', 'yamaha', 'suzuki']
print(motorcycles)
del motorcycles[0]
print(motorcycles)

['honda', 'yamaha', 'suzuki']
['yamaha', 'suzuki']


#### 🤓 List Comprehensions

sind eine elegante (pythonische) Methode zur Erzeugung numerischer Listen.

Eine List Comprehension kombiniert die for-Schleife und die Erzeugung neuer Elemente in einer Zeile und hängt automatisch fügt jedes neue Element an. List Comprehensions werden Anfängern nicht immer vorgestellt, da sie etwas ungewöhnlich aussehen. Allerdings werden Sie ihnen sicher bald im Code anderer Programmierer begegnen.

🤓✍️ **Aufgabe**
Schreiben Sie das folgende List Comprehension in eine reguläre Form mit einem for-loop um.


In [ ]:
squares = []
for value in range(1,11):
  squares.append(value**2)
print(squares)

In [ ]:
squares = [value**2 for value in range(1,11)]
print(squares)

🤓 **Inserting**

Sie können ein neues Element an einer beliebigen Position in Ihrer Liste hinzufügen, indem Sie die ```insert()```-Methode. Sie tun dies, indem Sie den Index des neuen Elements und den Wert des neuen Elements.

In [ ]:
motorcycles = ['honda', 'yamaha', 'suzuki']
motorcycles.insert(0, 'ducati')
print(motorcycles)

['ducati', 'honda', 'yamaha', 'suzuki']


🤓 **Popping**

Manchmal werden Sie den Wert eines Elements verwenden wollen, nachdem Sie es aus einer Liste entfernt haben. Zum Beispiel könnten Sie die Daten eines Experiments von allen Experimenten, die Sie auswerten müssen, erhalten wollen. 

Die `pop()`-Methode entfernt das letzte Element in einer Liste, aber sie lässt Sie mit diesem Element arbeiten, nachdem es entfernt wurde.

In [ ]:
lab_dates = ['21.03.2023', '22.03.2023', '24.03.2023']
print(lab_dates)
popped_lab_date = lab_dates.pop()
print(lab_dates)
print(popped_lab_date)

['21.03.2023', '22.03.2023', '24.03.2023']
['21.03.2023', '22.03.2023']
24.03.2023


Ähnlich wie `del` funktioniert auch `pop()` für jeden Index in der Liste.

In [ ]:
lab_dates = ['21.03.2023', '22.03.2023', '24.03.2023']
print(lab_dates)
popped_lab_dates = lab_dates.pop(1)
print(lab_dates)
print(popped_lab_dates)

['21.03.2023', '22.03.2023', '24.03.2023']
['21.03.2023', '24.03.2023']
22.03.2023


# 🤓 Tupel

Listen eignen sich gut für die Speicherung von Elementgruppen, die sich während der Laufzeit eines Programms ändern können. Die Möglichkeit, Listen zu ändern, ist besonders wichtig, wenn Sie mit einer Liste von Benutzern auf einer Website oder einer Liste von Figuren in einem Spiel arbeiten. 

Manchmal möchten Sie aber auch eine Liste von Elementen erstellen, die sich nicht ändern können. Mit Tupeln können Sie genau das tun. Python bezeichnet Werte, die sich nicht ändern können, als unveränderlich, und eine unveränderliche Liste wird als Tupel bezeichnet.

Tupel sind durch   `()` anstelle von `[]` gekennzeichnet und funktionieren sonst wie Listen. 

In [ ]:
dimensions = (200, 50)
print(dimensions[0])
print(dimensions[1])

200
50


versuchen, einen Wert innerhalb eines Tupels zu ändern:

In [ ]:
dimensions[0] = 250

TypeError: ignored

Wir können ein Tupel ändern, indem wir es umdefinieren:

In [ ]:
dimensions = (300, 25)
print(dimensions[0])
print(dimensions[1])

300
25


✍️🤓 **Aufgabe**

Sie erhalten die Daten der folgenden DNA-Sequenz.

```dna_sequence = ["A", "U", "C", "C", "G", "A", "G", "C", "U", "E", "G","A", "G", "C", "U", "Z", "G","A", "G", "C", "U",]```

Wie Sie sehen, müssen die Daten einige Fehler aufweisen, da Sie nur die Nukleotide A, U, C und G enthalten sollten. Bereinigen Sie zunächst die Daten, indem Sie alle fehlerhaften Elemente aus der Liste entfernen. Sie können die Funktion `remove()` verwenden oder nach einer eleganteren Methode googeln. Die `remove()` Methode entfernt jeweils nur den ersten gefundenen Wert von einer Liste.

Eine Sequenz von drei Nukleotiden wird als **Codon** bezeichnet, das in eine bestimmte Aminosäure übersetzt werden kann. 

Als nächstes erstellen Sie eine Schleife, die nacheinander jeweils in einer Zeile die drei Nukleotid-Buchstaben aller in der Liste enthaltenden Codons ausgibt.

![](https://www.genome.gov/sites/default/files/media/images/tg/Genetic-code.jpg)

In [ ]:
length_of_segments = 3
dna_sequence = ["A", "U", "C", "C", "G","A", "G", "C", "U", "E", "G","A", "G", "C", "U", "E", "Z", "G","A", "G", "C", "U","U"]

dna_sequence.remove("E")
dna_sequence.remove("Z")

for starting_point in range(0,len(dna_sequence), length_of_segments):
  
  print(dna_sequence[starting_point:starting_point+3])
  

['A', 'U', 'C']
['C', 'G', 'A']
['G', 'C', 'U']
['G', 'A', 'G']
['C', 'U', 'E']
['G', 'A', 'G']
['C', 'U', 'U']


🤓 **Bonus Aufgabe**:

Finden Sie einen Weg auch Listen mit mehreren falschen Werten auszuräumen indem Sie alle Z und E entfernen.

`dna_sequence = ["A", "U", "C", "C", "G","A", "G", "C", "U", "E", "G","A", "G", "C", "U", "E", "Z", "G","A", "G", "C", "U","U"]`



In [ ]:
length_of_segments = 3
dna_sequence = ["A", "U", "C", "C", "G","A", "G", "C", "U", "E", "G","A", "G", "C", "U", "E", "Z", "G","A", "G", "C", "U","U"]

#sublist created with list comprehension
dna_sequence_updated = [value for value in dna_sequence if (value != "E")]
dna_sequence_updated = [value for value in dna_sequence_updated if (value != "Z")]

#print original list
print(dna_sequence)

#print  new sublist that doesn't contain 'Python'
print(dna_sequence_updated)

✍️ 🤓 **Aufgabe**

Die Welt sehen: Denken Sie an mindestens fünf Orte auf der Welt, die Sie gerne
besuchen.
- 🤓 Speichern Sie die Orte in einer Liste. Achten Sie darauf, dass die Liste nicht in alphabetischer Reihenfolge ist.
- 🤓 Drucken Sie Ihre Liste in ihrer ursprünglichen Reihenfolge aus. Kümmern Sie sich nicht darum, die Liste ordentlich zu drucken, drucken Sie sie einfach als rohe Python-Liste aus.
- 🤓 Verwenden Sie `.sort()`, um die Liste in alphabetischer Reihenfolge zu drucken
- 🤓 Geben Sie das vorletzte Element aus
- 🤓 Geben Sie die Länge der Liste an
- 🤓 Fügen Sie ein weiteres Element hinzu
Wenn sie schneller sind, können Sie sich an folgenden Zusatzaufgaben versuchen:
- 🤓 Zeigen Sie, dass Ihre Liste immer noch in ihrer ursprünglichen Reihenfolge ist, indem Sie sie ausdrucken.
- 🤓 Verwenden Sie sorted(), um Ihre Liste in umgekehrter alphabetischer Reihenfolge zu drucken, ohne die
ohne die Reihenfolge der ursprünglichen Liste zu ändern.
- 🤓 Zeigen Sie, dass Ihre Liste immer noch in der ursprünglichen Reihenfolge ist, indem Sie sie erneut ausdrucken.
- 🤓 Verwenden Sie reverse=True, um die Reihenfolge der Liste zu ändern. Drucken Sie die Liste aus, um zu zeigen, dass ihre
Reihenfolge geändert hat.
- 🤓Verwenden Sie reverse=True, um die Reihenfolge der Liste erneut zu ändern. Drucken Sie die Liste aus, um zu zeigen
- 🤓Verwenden Sie sort(), um die Liste so zu ändern, dass sie in alphabetischer Reihenfolge gespeichert ist. Drucken Sie die
Liste aus, um zu zeigen, dass die Reihenfolge geändert wurde.
- 🤓Verwenden Sie sort(), um die Liste in umgekehrter alphabetischer Reihenfolge zu speichern.

Drucken Sie die Liste aus, um zu zeigen, dass die Reihenfolge geändert wurde.

In [ ]:
#@title ❓ Click `Show code` in the code cell to show the solution. { display-mode: "form" }

#1
list_of_places = ["Tirol", "Burgenland", "Wachau", "Finale Ligure", "Provence"]

#2
print(list_of_places)

#3
list_of_places.sort()
print(list_of_places)

#4
print(list_of_places[-2])

#5
print(len(list_of_places))

#6
list_of_places.append("Innsbruck")

['Tirol', 'Burgenland', 'Wachau', 'Finale Ligure', 'Provence']
['Burgenland', 'Finale Ligure', 'Provence', 'Tirol', 'Wachau']
Tirol
5


In [ ]:
#@title ❓ Click `Show code` in the code cell to show the solution. { display-mode: "form" }

#1
list_of_places = ["Tirol", "Burgenland", "Wachau", "Finale Ligure", "Provence"]

#2
print(list_of_places)

#3
print(sorted(list_of_places))

#4
print(list_of_places)

#5
print(sorted(list_of_places,reverse=False))

#6
print(list_of_places)

#7
print(sorted(list_of_places,reverse=True))

#8
print(list_of_places)

#9
list_of_places.sort(reverse=False)
print(list_of_places)

#10
list_of_places.sort(reverse=False)

['Tirol', 'Burgenland', 'Wachau', 'Finale Ligure', 'Provence']
['Burgenland', 'Finale Ligure', 'Provence', 'Tirol', 'Wachau']
['Tirol', 'Burgenland', 'Wachau', 'Finale Ligure', 'Provence']
['Burgenland', 'Finale Ligure', 'Provence', 'Tirol', 'Wachau']
['Tirol', 'Burgenland', 'Wachau', 'Finale Ligure', 'Provence']
['Wachau', 'Tirol', 'Provence', 'Finale Ligure', 'Burgenland']
['Tirol', 'Burgenland', 'Wachau', 'Finale Ligure', 'Provence']
['Burgenland', 'Finale Ligure', 'Provence', 'Tirol', 'Wachau']
